CRM/MembershipExtras/Form/RecurringContribution/AddNoInstalmentsMembershipLineItem.php

```diff
@@ -0,0 +1,296 @@
+  private function calculateProRataDaysCount() {
+    $membershipStartDate = new DateTime($this->lineItemParams['start_date']);
+    $membershipEndDate = new DateTime($this->lineItemParams['end_date']);
+    return $membershipEndDate->diff($membershipStartDate)->format("%a");
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function buildQuickForm() {
+    CRM_Utils_System::setTitle(E::ts('Add %1', ['1' => $this->membershipType->name]));
+
+    $this->addRadio('payment_type', '', [
+      self::PAYMENT_TYPE_NO_PAYMENT => ts('No Payment'),
+      self::PAYMENT_TYPE_ONE_OFF_PAYMENT => ts('One off payment'),
+    ]);
+
+    $this->add('datepicker', 'scheduled_charge_date', ts('Scheduled Charge Date'), [], TRUE, ['time' => FALSE]);
+
+    $this->addMoney('amount_exc_tax', ts('Amount exc Tax'), TRUE, [], FALSE);
+
+    $financialTypes = CRM_Financial_BAO_FinancialType::getAvailableFinancialTypes();
+    $this->add('select', 'noinstalmentline_financial_type_id', ts('Financial Type'), $financialTypes, TRUE);
+
+    $this->addMoney('amount_inc_tax', ts('Amount inc Tax'), FALSE, ['readonly' => TRUE], FALSE);
+
+    $this->add('checkbox', 'noinstalmentline_send_confirmation_email', ts('Send confirmation email?'));
+
+    $this->assign('prorataDaysCount', $this->prorataDaysCount);
+
+    $this->addButtons([
+      [
+        'type' => 'submit',
+        'name' => E::ts('Apply'),
+        'isDefault' => TRUE,
+      ],
+      [
+        'type' => 'cancel',
+        'name' => E::ts('Cancel'),
+        'isDefault' => FALSE,
+      ],
+    ]);
+  }
+
+  public function setDefaultValues() {
+    return [
+      'payment_type' => 1,
+      'scheduled_charge_date' => date('Y-m-d'),
+      'noinstalmentline_financial_type_id' => $this->membershipType->financial_type_id,
+      'amount_exc_tax' => $this->calculateAmountExcTax(),
+    ];
+  }
+
+  private function calculateAmountExcTax() {
+    $membershipTypeDatesCalculator = new CRM_MembershipExtras_Service_MembershipTypeDatesCalculator();
+    $membershipDurationCalculator = new CRM_MembershipExtras_Service_MembershipTypeDurationCalculator($this->membershipType, $membershipTypeDatesCalculator);
+
+    $proratedAmount = $this->membershipType->minimum_fee;
+    if ($this->prorataDaysCount > 1) {
+      $membershipTypeDurationInDays = $membershipDurationCalculator->calculateOriginalInDays();
+      $proratedAmount = ($this->membershipType->minimum_fee / $membershipTypeDurationInDays) * $this->prorataDaysCount;
+      $proratedAmount = CRM_MembershipExtras_Service_MoneyUtilities::roundToPrecision($proratedAmount, 2);
+    }
+
+    return $proratedAmount;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function postProcess() {
+    $this->submittedValues = $this->exportValues();
+
+    $transaction = new CRM_Core_Transaction();
+    try {
+      $this->processPostRequest();
+      $transaction->commit();
+    }
+    catch (Exception $e) {
+      $transaction->rollback();
+      $this->showErrorNotification($e);
```

Comment:
Should we log the error too? So, we make it easier to investigate?

Reply:
I don't think it is important, given if there is a problem then it will be rolled back, and then the user who got the error can just share the steps they did with us and we should be able to replicate.